### DISCUSSION TITLE
Component initialisation

### DISCUSSION THREAD

**Discussion by hchockarprasad** [ivanceras/sauron#79] [Q&A] [answered]
Component init method doesn’t seems to be triggered…Is this a bug or should we  need to call manually..If so where would be the best place to call manually?

**Comment by ivanceras** [comment, marked as answer]
@hchockarprasad, You have to manually call your Component's `init` function from your Application. Just like `update` in Component is also wired through your Application's `update` function. I haven't included a simplified example for this use-case, since I'm still stabilizing the method signatures of this `init` function in the Application and in the Component. However, A more advance usage is demonstrated here in the [ultron editor](https://github.com/ivanceras/ultron), which I also wrote.

From the App's [init function](https://github.com/ivanceras/ultron/blob/6a58fa020e6ecb4ded286527e075cc3c53386017/packages/ultron-app/src/app.rs#L47)

```rust
    fn init(&mut self) -> Cmd<Self, Msg> {
        Cmd::batch([
            Cmd::new(|program| {
                program.add_window_event_listeners(vec![
                    on_mousemove(|me| Msg::WebEditorMsg(web_editor::Msg::Mousemove(me))),
                    on_mousedown(|me| Msg::WebEditorMsg(web_editor::Msg::Mousedown(me))),
                    on_mouseup(|me| Msg::WebEditorMsg(web_editor::Msg::Mouseup(me))),
                ])
            }),
            Cmd::from(self.web_editor.init().localize(Msg::WebEditorMsg)),
        ])
    }

```
It calls into the [`init`](https://github.com/ivanceras/ultron/blob/6a58fa020e6ecb4ded286527e075cc3c53386017/packages/ultron-web/src/web_editor.rs#L188) function of the WebEditor Component, which also then calls to `init` function of its sub components. 

In summary, Application is the only interface where the `init` function is called automatically, and it's up for you to wire its components and sub-components. The `Program` object only expects to be passed with an `Application` object. The `Component` is there for Composability and Convenience, to be used by your `App`. The program has no knowledge about `Components`.
